_posts/2020-04-28-mocking.adoc

```diff
@@ -0,0 +1,375 @@
+</dependency>
+----
+
+Now we can rewrite the `MockTestCase` like so:
+
+[source,java]
+----
+@QuarkusTest
+public class MockTestCase {
+
+    @InjectMock
+    MockableBean1 mockableBean1; // <1>
+
+    @InjectMock
+    MockableBean2 mockableBean2;
+
+    @BeforeEach
+    public void setup() {
+        Mockito.when(mockableBean1.greet("Stuart")).thenReturn("A mock for Stuart"); // <2>
+    }
+
+    @Test
+    public void firstTest() {
+        Assertions.assertEquals("A mock for Stuart", mockableBean1.greet("Stuart"));
+        Assertions.assertEquals(null, mockableBean2.greet("Stuart"));
+    }
+
+    @Test
+    public void secondTest() {
+        Mockito.when(mockableBean2.greet("Stuart")).thenReturn("Bonjour Stuart"); // <3>
+        Assertions.assertEquals("A mock for Stuart", mockableBean1.greet("Stuart"));
+        Assertions.assertEquals("Bonjour Stuart", mockableBean2.greet("Stuart"));
+    }
+
+    @ApplicationScoped
+    public static class MockableBean1 {
+
+        public String greet(String name) {
+            return "Hello " + name;
+        }
+    }
+
+    @ApplicationScoped
+    public static class MockableBean2 {
+
+        public String greet(String name) {
+            return "Hello " + name;
+        }
+    }
+}
+----
+<1> `@InjectMock` results in a mock being created and being available in *all* test methods of the test class (other test classes are *not* affected by this)
+<2> The `mockableBean1` is configured here for all test methods of the class
+<3> Configure `mockableBean2` for this test only
+
+[NOTE]
+====
+Since `@InjectMock` uses `QuarkusMock` under the hood, the same limitations apply to its use.
+
+Additionally, `@InjectMock` works like an injection point for the bean, so for it to work properly when
+the target bean uses CDI qualifiers, those qualifiers also need to be added to the field.
+We will see an example of this in next section about mocking a `@RestClient` bean.
+====
+
+As a final example, we can rewrite the `OrderServiceTest` test like so:
+
+[source,java]
+----
+public class OrderServiceTest {
+
+    @Inject
+    private OrderService orderService;
+
+    @InjectMock
+    private InvoiceNotificationService invoiceNotificationService;
+
+    @BeforeAll
+    public static void setup() {
+        doNothing().when(invoiceNotificationService).sendInvoice(any());
+        doNothing().when(invoiceNotificationService).notifyInvoiceAlreadySent(any());
+    }
+
+    public void testGenerateSendInvoice() {
+        // perform some setup
+
+        Invoice invoice = orderService.generateSendInvoice(1L);
+
+        // perform some assertions
+    }
+}
+----
+
+==== Using @InjectMock with @RestClient
+
+A very common need is to mock `@RestClient` beans. Thankfully it's a need well covered by `@InjectMock` - as long
+as two things principles are followed:
+
+* The bean is made `@ApplicationScoped` (instead of accepting the default scope which `@RegisterRestClient` implies, i.e. `@Singleton`)
```

Comment:
Shouldn't we change the default scope for rest clients then? Hm, looking at the spec it should be `@Dependent` by default. And it seems that we use `@Dependent` by default...

Reply:
You are right, it is `@Dependent`

Reply:
Fixed